src/gui/dialogs/statistics_dialog.cpp

```diff
@@ -164,19 +183,83 @@ void statistics_dialog::add_damage_row(
 // Return the string to use in the "Hits" table, showing actual and expected number of hits.
 static std::string tally(const std::map<int, struct statistics::stats::by_cth_t>& by_cth)
 {
-	int overall_hits = 0;
+	unsigned int overall_hits = 0;
 	double expected_hits = 0;
+	unsigned int overall_strikes = 0;
 
 	for(const auto& i : by_cth) {
 		int cth = i.first;
 		overall_hits += i.second.hits;
 		expected_hits += (cth * 0.01) * i.second.strikes;
+		overall_strikes += i.second.strikes;
 	}
 
-	// TODO: show a priori probability of this actual result here. https://en.wikipedia.org/wiki/Poisson_binomial_distribution
-
 	std::ostringstream str;
 	str << overall_hits << " / " << expected_hits;
+
+	// Compute the a priori probability of this actual result, by simulating many attacks against a single defender.
```

Comment:
> It's even more confusing that depending on the number of hits, the percentage may the probability of _this exact result_, the probability of _this or more hits_, or the probability of _this or less hits_, i.e. completely different things.

"This exact result" is a rare case since the expected value is usually not an integer. Perhaps we could display an em dash in that case? There is precedent for that: the "Gold cost" column in the "Advancements" row shows an em dash.

The other two cases are exactly the same: they show the _a priori_ probability of the actual result. Isn't that what players would want to know? When a player hits above average, they would ask "How lucky was I"; when a player hits below average, they would ask "How unlucky was I".

There is a tooltip that explains this briefly: https://github.com/wesnoth/wesnoth/pull/4070/commits/67847506d4c1d1dae1eb19f820fea02c2fb6e0c6#diff-6f98c6223f5ee298110c1deef256831eR562

I suspected a tooltip wouldn't be enough but wasn't sure what else to do. I suppose I could add a static label to the dialog, or a new page to the in-game help section, to explain what the table shows? Would that address the issue?

What do you think of showing "probability of having this many hits or more"? This would make it so if a player was very _lucky_ the percentage shown would be small, but if the player was very _unlucky_ the percentage shown would be large. I honestly don't know if I like this better than the currently-implemented behavior. Treating the two extreme cases differently like this feels wrong somehow. If a unit has 50% chance to hit, then "very lucky" and "very unlucky" are _both_ rare/extreme cases, so showing a small percentage in both of these cases feels appropriate.

Reply:
I think always showing "probability of this many hits or more" would be significantly more consistent, and thus a huge improvement.

Reply:
:thinking: 

I see your point that showing different things in the same label is poor UI design, but I stand by my point that "very lucky" and "very unlucky" should be presented similarly. For example, there is no particular reason to prefer implementing "always show 'probability of this many hits or more'" over implementing "always show 'probability of this many hits or _less_'".

That being said, I'm not sure what change to make. I just [asked the forums](https://forums.wesnoth.org/viewtopic.php?f=12&t=49785&p=642368#p642368) for ideas. I'd like to address @Pentarctagon's feedback at the same time.

Reply:
I've just pushed two commits (e70c1ed863ee4565689ee174ff4224354bf0ee65 and 1dfe728e04caea865ae17ceb998051da8da58957). The log messages explain what they do. Here's what they look like:

![2019-05-19-223318_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/57989263-cf622480-7a87-11e9-815c-48d535f243b7.png)
![2019-05-19-223411_1920x1080_scrot](https://user-images.githubusercontent.com/24784687/57989264-cf622480-7a87-11e9-9e38-d434d1f67f0b.png)

The screenshots show (especially in the "Overall" column) how when the RNG was on your side the numbers are red and when the RNG was on the enemy's side the numbers are green.

Reply:
I think it would work better with some simplification. You could remove the "probability of actual result" and the coloring, they're both redundant.

Reply:
Thanks for the suggestions. I'll remove "probability of actual result". (I might patch it back in in my local branch, though.) As to coloring, I think it helps, because whether large numbers are good or bad varies. In the screenshot, 63.9% and 71.5% are colored in "green to red", and 24.6% and 18.4% are colored in "red to green". For example, if the stats were 5%/0%/95% in both taken and inflicted, the first row would be green and the second one red. I think it would be poor UI to just show all four of these figures in white.

On the other hand, I decided against coloring the percentages in the damage table, because in that case I do think coloring is redundant. Because of regression to the mean, the percentage values in that table are always around 0%, and when they aren't, the sample size is too small anyway.